Neos.Neos/Classes/Domain/Service/WorkspaceService.php

```diff
@@ -0,0 +1,365 @@
+            'title' => $newWorkspaceTitle->value,
+            'description' => $newWorkspaceDescription->value,
+        ]);
+    }
+
+    /**
+     * Retrieve the currently active personal workspace for the specified $userId
+     *
+     * NOTE: Currently there can only ever be a single personal workspace per user. But this API already prepares support for multiple personal workspaces per user
+     */
+    public function getPersonalWorkspaceForUser(ContentRepositoryId $contentRepositoryId, UserId $userId): Workspace
+    {
+        $workspaceName = $this->findPrimaryWorkspaceNameForUser($contentRepositoryId, $userId);
+        if ($workspaceName === null) {
+            throw new \RuntimeException(sprintf('No workspace is assigned to the user with id "%s")', $userId->value), 1718293801);
+        }
+        return $this->requireWorkspace($contentRepositoryId, $workspaceName);
+    }
+
+    /**
+     * Create a new root (aka base) workspace with the specified metadata
+     */
+    public function createRootWorkspace(ContentRepositoryId $contentRepositoryId, WorkspaceTitle $title, WorkspaceDescription $description): WorkspaceName
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+        $workspaceName = self::getUniqueWorkspaceName($contentRepository, $title->value);
+        $contentRepository->handle(
+            CreateRootWorkspace::create(
+                $workspaceName,
+                DeprecatedWorkspaceTitle::fromString($title->value),
+                DeprecatedWorkspaceDescription::fromString($description->value),
+                ContentStreamId::create()
+            )
+        );
+        $this->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::ROOT, null);
+        return $workspaceName;
+    }
+
+    /**
+     * Create a new, personal, workspace for the specified user
+     */
+    public function createPersonalWorkspace(ContentRepositoryId $contentRepositoryId, WorkspaceTitle $title, WorkspaceDescription $description, WorkspaceName $baseWorkspaceName, UserId $ownerId): WorkspaceName
+    {
+        return $this->createWorkspace($contentRepositoryId, $title, $description, $baseWorkspaceName, $ownerId, WorkspaceClassification::PERSONAL);
+    }
+
+    /**
+     * Create a new, potentially shared, workspace
+     */
+    public function createSharedWorkspace(ContentRepositoryId $contentRepositoryId, WorkspaceTitle $title, WorkspaceDescription $description, WorkspaceName $baseWorkspaceName, UserId|null $ownerId = null): WorkspaceName
+    {
+        return $this->createWorkspace($contentRepositoryId, $title, $description, $baseWorkspaceName, $ownerId, WorkspaceClassification::SHARED);
+    }
+
+    /**
+     * Create a new, personal, workspace for the specified user if none exists yet
+     */
+    public function createPersonalWorkspaceForUserIfMissing(ContentRepositoryId $contentRepositoryId, User $user): void
```

Comment:
the `void` signature here and the throwing at `getPersonalWorkspaceForUser` makes this not easily usable id say.

A future requirement of the Neos ui will be to get the content graph for the current user. For that we will have to figure out if a personal workspace already exists or create a new one based on live.

The code with the current api would be following:

```php
$service->createPersonalWorkspaceForUserIfMissing($cr, $user);
$workspace = $service->getPersonalWorkspaceForUser($cr, $user->getId());
$contentGraph = $cr->getContentGraph($workspace->workspaceName);
```

there are 3 things that i noticed:
- `$user->getId()` vs `$user`
- we need to fetch the workspace name via `findPrimaryWorkspaceNameForUser` twice and just discard the first result (could be resolved by returning it)
- we technically dont need to fetch the workspace at all but only its name to get the content graph (especially if fetching workspaces will be slightly more expensive)

Reply:
> $user->getId() vs $user

what about it?

> we technically dont need to fetch the workspace at all but only its name to get the content graph (especially if fetching workspaces will be slightly more expensive)

isn't that a micro optimization for some method that is almost never called?

Reply:
> what about it?

we should be consistent of either passing the whole user vs id inside one class i think ^^ but thats just a nitpick

> isn't that a micro optimization for some method that is almost never called?

dont worry i only expressed my concerns here because it is called on every call in the neos ui backend `neos/content` so ...

Reply:
> we should be consistent of either passing the whole user vs id inside one class i think

Ah, gotcha.
I pass the whole user object because we need the user's _name_ too, to create a speaking title (see https://github.com/neos/neos-development-collection/blob/88e344d2068a0ef451cde56725a7960013fe2a51/Neos.Neos/Classes/Domain/Service/WorkspaceService.php#L144)

> dont worry i only expressed my concerns here because it is called on every call in the neos ui backend

but the expensive part would only be executed if no primary workspace was found, right?!

Reply:
1. ahh i see sorry oversaw that ^^

2. yes the creation - the expensive part - only happens once, i was just thinking about how to optimise the content graph loading for when opening the backen but it probably doesnt matter and is a sidetrack. Premature optimisation etc... so lets go down the route as drawn out and see if we want to optimise it later ;)

Reply:
1. ahh i see sorry oversaw that ^^

2. yes the creation - the expensive part - only happens once, i was just thinking about how to optimise the content graph loading for when opening the backen but it probably doesnt matter and is a sidetrack. Premature optimisation etc... so lets go down the route as drawn out and see if we want to optimise it later ;)